### DISCUSSION TITLE
Remove build.format: 'file'

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#1003] [Proposal]
### Body

# Summary

A brief, one or two sentence explanation of the proposal.
Remove the `'file'` option from `build.format`, as the newer `'preserve'` option allows for the flexibility of defining the output file in an easier to understand way.

# Background & Motivation

The `'file'` option was initialized created in order for Astro to work with static servers that do not support the `dir/index.html` pattern. In these servers the URL is for a specific file and for some `/about.html` is perceived to be nicer than `about/index.html`.

Later it was discovered that, due to a bug, the `'file'` option would squash folders you have created yourself, such as `about/index.astro` and output `about.html`. This was not done intentionally.

When we went to fix this we realized it was a breaking change, and a new option `'preserve'` was created that matches the input filesystem structure exactly. This gives you the flexibility to have the output structure you want in an easy to understand way.

Given this, there's no real use-case for `'file'` so this proposal is to remove it.

# Goals

- Simplify internal codebase.
- Remove difficult to understand `'file'` option.

**Comment by wassfila** [comment]
This looks like an improvement to reduce things people need to think about. The formulation of this sentence is confusing me a bit "Remove the 'file' option from build.format, " so do we remove the 'file' and leave both 'directory' and 'preserve' as options ? which then would feels like an asymetric flexibility, or do we simply remove the whole "build.format" which is no longer configurable and frozen in 'preserve' option which matches the simplification argument, which I find good and fair.

**Reply by matthewp** [reply]
This doesn't propose removing `'directory'` which is what most people want / expect anyways. For a greenfield project on a modern host it gives you pretty URLs and you don't have to think about it. It becomes:

- `'directory'`: opinionated default that allows most people not to think about build output format.
- `'preserve'`: if you have legacy URLs you need to match or an unusual host, this gives you full flexibility.

**Reply by bluwy** [reply]
I agree with @wassfila that `file` and `directory` are counterparts to each other. There should be both if one exists, and this also reflects how `trailingSlash: 'never' | 'always'` works.

I'd say that `preserve` is the better default as it leaves the output as you define it, and `file` and `directory` are for those who prefer to consistently access a path with no trailing slash or with trailing slash consistently respectively. Not saying that we should change the default, but each has its own behaviour and usecase.

**Reply by matthewp** [reply]
I don't think symmetry is a great reason for having a feature. Just because it's the opposite of `directory` doesn't mean its useful, and I don't think that's really the case here. `directory` is a DX feature to allow you to add pages without thinking about the output structure in an opinionated (and normal) way. I don't think that's really true of `file`, it's pretty niche to even need it.

**Comment by delucis** [comment]
Noting that Starlight uses `build.format: 'file'` as a “proxy” for whether to render links with or without `.html`. But that’s always been a bit of a funny one, so while we will need to think of something (maybe nothing initially as we currently treat `'file'` and `'preserve'` as equivalent?), no opposition to this change from me.

**Comment by bluwy** [comment]
> Later it was discovered that, due to a bug, the `'file'` option would squash folders you have created yourself, such as `about/index.astro` and output `about.html`. This was not done intentionally.

I don't understand how this is a bug, isn't that's what `'file'` is for? To convert all directories/index.html to files. Otherwise what is the original intent?

**Reply by matthewp** [reply]
In my mind `file` was always to avoid _creating_ folders, because asking the user to type `http://example.com/about.html` is more acceptable than asking them to type `http://example/about/index.html`. If you want `about.html` then why are you creating a folder with an `index.astro`?

**Reply by jcayzac** [reply]
@matthewp because I may also have an index.png.ts and an index.xml.ts in that same folder and I don't want to repeat the parameters in the folder's name, e.g. `v[video]`.

**Reply by bluwy** [reply]
Example: let's say I have a `src/pages/blog/` folder which contains some blog posts, and I want an `index.html` to list all the blog post. In this case, you'd now get `src/pages/blog/index.html` instead of `src/pages/blog.html` (and it'd be weird if there's both the `blog` folder and `blog.html` file).

So I think it's common to have structures like this that also mean differently, and some may prefer in this case to use `'file'` to make sure all of them are files and can access them without a trailing slash.

**Comment by matthewp** [comment]
Dropping this for now. I wasn't expecting as much push back against this. Good to know.
